pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,355 @@
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+
+	cnsvsphere "sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/logger"
+)
+
+const (
+	waitForUpdatesRetry = 5 * time.Second
+	infoPropertyName    = "info"
+)
+
+type Result struct {
+	TaskInfo *types.TaskInfo
+	Err      error
+}
+
+// TaskDetails is used to hold state for a task
+// param: MarkedForRemoval helps in retrying the removal of tasks in case of failures
+type TaskDetails struct {
+	Reference        types.ManagedObjectReference
+	MarkedForRemoval bool
+}
+
+// RemoveTasksMarkedForDeletion goes over the list of tasks in the map
+// and removes tasks that have been marked for deletion
+func (t *TaskViewManagerImpl) RemoveTasksMarkedForDeletion() {
+	ctx := context.Background()
+	log := logger.GetLogger(ctx)
+	// todo: do we need a mutex here if we are accessing items that others won't anymore?
+	listView := t.getListView()
+	if listView == nil {
+		log.Errorf("ListView is empty. Will attempt to remove invalid tasks in next attempt. ")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	log.Infof("pending tasks count before purging: %v", len(t.taskToChannelMap))
+	for taskDetails := range t.taskToChannelMap {
+		if taskDetails.MarkedForRemoval == true {
+			err := listView.Remove(ctx, []types.ManagedObjectReference{taskDetails.Reference})
+			if err != nil {
+				log.Errorf("failed to remove task from ListView. error: %+v", err)
+				continue
+			}
+			delete(t.taskToChannelMap, taskDetails)
+		}
+	}
+	log.Infof("pending tasks count after purging: %v", len(t.taskToChannelMap))
+}
+
+// MarkForDeletion marks a given task MoRef for deletion by setting a boolean flag in the TaskDetails object
+func (t *TaskViewManagerImpl) MarkForDeletion(ctx context.Context, taskMoRef types.ManagedObjectReference) error {
+	log := logger.GetLogger(ctx)
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	ch, ok := t.taskToChannelMap[TaskDetails{Reference: taskMoRef}]
+	if !ok {
+		return logger.LogNewErrorf(log, "failed to retrieve receiver channel for task %+v", taskMoRef)
+	}
+	delete(t.taskToChannelMap, TaskDetails{Reference: taskMoRef})
+	t.taskToChannelMap[TaskDetails{Reference: taskMoRef, MarkedForRemoval: true}] = ch
+	return nil
+}
+
+func (t *TaskViewManagerImpl) GetPendingTasksCount() int {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return len(t.taskToChannelMap)
+}
+
+func NewTaskViewManager(virtualCenter *cnsvsphere.VirtualCenter) *TaskViewManagerImpl {
+	ctx := context.Background()
+	m := make(map[TaskDetails]chan Result)
+	t := &TaskViewManagerImpl{
+		virtualCenter:    virtualCenter,
+		ctx:              ctx,
+		taskToChannelMap: m,
+	}
+	go t.init(t.ctx)
+	return t
+}
+
+// TaskViewManagerImpl is the struct used to manage a single listView instance.
+// params
+// virtualCenter: holds a reference to the global VC object
```

Comment:
This seems to be designed only for single vc setup, noticed that the listView is also associated with a single vc. Can we do something to ensure this is really compatible with multi-vc setups? you will need to do this sooner or later.